### DISCUSSION TITLE
Typing relationship with a ORM mapped class defined in another module

### DISCUSSION THREAD

**Discussion by DiegoPiloni** [sqlalchemy/sqlalchemy#9522] [Usage Questions] [answered]
### Ensure stubs packages are not installed

- [X] No sqlalchemy stub packages is installed (both `sqlalchemy-stubs` and `sqlalchemy2-stubs` are not compatible with v2)

### Verify if the api is typed

- [X] The api is not in a module listed in [#6810](https://github.com/sqlalchemy/sqlalchemy/issues/6810) so it should pass type checking

### Describe the typing issue

Hi ! First of all, thanks for version 2 :muscle:  and excuse my bad english.

I'm defining both ORM mapped classes showed in the [SQLAlchemy Unified Tutorial](https://docs.sqlalchemy.org/en/20/tutorial/metadata.html#declaring-mapped-classes) each one in its own module: `user.py`  and `address.py`. 

If I try to refer them as strings in relationships typing, running `mypy` reports an error that the names are not defined.

Is this something that could be solved in some way? Or should mapped classes be all defined in the same module to refer to each other in relationships typing  ?

### To Reproduce

```python
from typing import List
from typing import Optional
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column
from sqlalchemy.orm import relationship

# user.py
class User(Base):
    __tablename__ = "user_account"
    id: Mapped[int] = mapped_column(primary_key=True)
    name: Mapped[str] = mapped_column(String(30))
    fullname: Mapped[Optional[str]]
    addresses: Mapped[List["Address"]] = relationship(back_populates="user")
    def __repr__(self) -> str:
        return f"User(id={self.id!r}, name={self.name!r}, fullname={self.fullname!r})"

# address.py
class Address(Base):
    __tablename__ = "address"
    id: Mapped[int] = mapped_column(primary_key=True)
    email_address: Mapped[str]
    user_id = mapped_column(ForeignKey("user_account.id"))
    user: Mapped["User"] = relationship(back_populates="addresses")
    def __repr__(self) -> str:
        return f"Address(id={self.id!r}, email_address={self.email_address!r})"
```

### Error

```
app/models/user.py:X: error: Name "Address" is not defined  [name-defined]
app/models/address.py:Y: error: Name "User" is not defined  [name-defined]
```

### Versions

- OS: Ubuntu 21.10
- Python: 3.10.10
- SQLAlchemy: 2.0.7
- Type checker (eg: mypy 0.991, pyright 1.1.290, etc): mypy 1.1.1

**Comment by CaselIT** [comment, marked as answer]
Hi,

Python does not support circular imports, so to solve this for typing purposes (that includes many more cases where circular imports are required) the std lib has [`TYPE_CHECKING`](https://docs.python.org/3/library/typing.html#typing.TYPE_CHECKING)

So you can do:
```py
# in user
from typing import TYPE_CHECKING
if TYPE_CHECKING:
  from .address import Address

# in address instead
from typing import TYPE_CHECKING
if TYPE_CHECKING:
  from .user import User
```

note that this is a general python typing practice, it's not something custom of sqlalchemy so you can use it in any python program
also `Address` or `User` are not actually imported, so using them directly (like trying to create an instance of user in address.py) will cause an error

**Reply by DiegoPiloni** [reply]
Thanks a lot @CaselIT ! This works as expected :muscle:

**Reply by zzzeek** [reply]
I've answered your question at https://github.com/sqlalchemy/sqlalchemy/discussions/9649#discussioncomment-5619928
